opencog/atoms/core/ImplicationLink.cc

```diff
@@ -61,26 +61,23 @@ ImplicationLink::ImplicationLink(Link &l)
 void ImplicationLink::extract_variables(const HandleSeq& oset)
 {
 	size_t sz = _outgoing.size();
-	// For now we assume no implicit scoping, so the number of
-	// arguments is 2 when sugar syntax isn't used (no scoping), or 3
-	// when the sugar syntax (scoping) is used.
+
 	if (sz < 2 or 3 < sz)
 		throw InvalidParamException(TRACE_INFO,
-			"Expecting an outgoing set size of at most two, got %d", sz);
+			"Expecting an outgoing set of size 2 or 3, got %d", sz);
+
+	// If there is an explicit variable declaration, then use that.
+	// If there is no variable declaration, then assume that all of
+	// the free variables in the link are scoped.
+	ScopeLink::extract_variables(oset);
```

Comment:
When I revert that changes I get numerous unit tests that fail, that scares me a bit

```
The following tests FAILED:
	 14 - MultiSpaceUTest (Failed)
	 18 - MapLinkUTest (Failed)
	 30 - FindUtilsUTest (Failed)
	 41 - PatternCrashUTest (Failed)
	 58 - RecognizerUTest (Failed)
	 59 - ArcanaUTest (Failed)
	 61 - GetLinkUTest (Failed)
	 66 - EvaluationUTest (Failed)
	 67 - QuoteUTest (Failed)
	 70 - BuggyQuoteUTest (Failed)
	 82 - TypeChoiceUTest (Failed)
	 85 - UREConfigReaderUTest (Failed)
	 86 - ForwardChainerUTest (Failed)
	 87 - BackwardChainerUTest (Failed)
```

Any idea?